File: client/src/protocol/handshake.zig
const std = @import("std");
const protocol = @import("protocol.zig");

pub fn build(group: []const u8, alloc: std.mem.Allocator) ![]u8 {
    const data = protocol.Data{
        .data_type = protocol.DataType.handshake,
        .body = group,
    };

    return data.encode(alloc);
}

File: server/protocol/handshake.go
package protocol

import "fmt"

func Handshake(data []uint8) (string, []byte, error) {
	d, err := Decode(data)
	if err != nil {
		return "", nil, err
	}

	if d.DataType != typeHandshake {
		return "", nil, fmt.Errorf("invalid data type: %d", d.DataType)
	}

	group := string(d.Body)
	if group == "" {
		group = "unknown"
	}

	return group, buildOK(typeHandshake), nil
}
